open-apparel-registry/open-apparel-registry
# Add OAR ID generation

## Overview

This commit implements a new process for generating OAR IDs

Connects #276

### Notes

The new IDs have the following structure:

Position   Value
1-2        Country code
3-9        When OAR ID was generated as 4-digit year and 3-digit day of the year
10-14       A random value encoded from a restricted character set
15         A check digit

An example, with optional hyphens separating the tokens:

CN-2019067-NZ95A-M

This schema has the following desirable qualities:

- We don't know much about a facility when it is added to the Facility table. Country code and the date the ID was created are two useful, immutable pieces of information that we can use to build a meaningful ID.
- The random portion of the ID inherits these desirable qualities from the base32 implementation used to create them (https://www.crockford.com/base32.html):
  - Human readable and machine readable.
  - Compact. Humans have difficulty in manipulating long strings of arbitrary symbols.
  - Error resistant. Entering the symbols must not require keyboarding gymnastics.
  - Pronounceable. Humans should be able to accurately transmit the symbols to other humans using a telephone.
- The check digit allows the value to be algorithmically verified for errors.
- It is shorter than the GUID values used in the previous version of the
  application.

A 5-digit base32 number can encode around 32 million values. This combined with the date and country means that we are unlikely to exhaust the pool of available IDs.

## Testing Instructions

* `./scripts/update` to install the new Python dependency
* Verify that the unit test pass
* Run `manage.py resetdb` and verify that the fixtures load correctly.
* Register and log in
* Upload  ⁨src⁩/django⁩/api⁩/management⁩/commands⁩/facility_lists⁩/13.csv
* Fully process the CSV
  * `./scripts/manage batch_process --list-id 15 --action parse`
  * `./scripts/manage batch_process --list-id 15 --action geocode`
  * `./scripts/manage batch_process --list-id 15 --action match`
* Verify that new facilities are created with new IDs
